text/0000-build-systems.md

```diff
@@ -0,0 +1,455 @@
+into a larger build system. This finer division is left as a question for
+experimentation.
+
+## Specifics for the homogeneous build system case
+
+For homogeneous build systems, there are two kinds of code that must be dealt
+with: code originally written using vanilla Cargo and a crate registry, and code
+written "natively" in the context of the external build system.  Any integration
+has to handle the first case (to have access to crates.io or a vendored mirror
+thereof).
+
+### Using crates vendored from or managed by a crate registry
+
+Whether using a registry server or a vendored copy, if you're building Rust code
+that is written using vanilla Cargo, you will at some level need to use Cargo's
+dependency resolution and `Cargo.toml` files. In this case, the external build
+system should invoke Cargo for *at least* the dependency resolution and build
+configuration steps, and likely the build lowering step as well. In such a
+world, Cargo is responsible for *planning* the build (which involves largely
+Rust-specific concerns), but the external build system is responsible for
+*executing* it.
+
+A typical pattern of usage is to have a whitelist of "root dependencies" from an
+external registry which will be permitted as dependencies within the
+organization, often pinning to a specific version and set of Cargo
+features. This whitelist can be described as a single `Cargo.toml` file, which
+can then drive Cargo's dependency resolution just once for the entire registry.
+The resulting lockfile can be used to guide vendoring and construction of a
+build plan for consumption by the external build system.
+
+One important concern is: how do you depend on code from other languages, which
+is being managed by the external build system? That's a narrow version of a more
+general question around *native dependencies*, which will be addressed
+separately in a later section.
+
+#### Workflow and interop story
+
+On the external build system side, a rule or plugin will need to be written that
+knows how to invoke Cargo to produce a build plan corresponding to a whitelisted
+(and potentially vendored) registry, then translate that build plan back into
+appropriate rules for the build system. Thus, when doing normal builds, the
+external build system drives the entire process, but invokes Cargo for guidance
+during the planning stage.
+
+### Using crates managed by the build system
+
+Many organization want to employ their own strategy for maintaining and
+versioning code, and for resolving dependencies. In this case, they may wish to
+entirely forgo producing a meaningful Cargo.toml for the code the write, instead
+having one that just forwards to a plugin. The description of dependencies is
+then written in the external build system's rule format. Here, Cargo acts
+primarily as a *workflow and tool orchestrator*, since it is not involved in
+either planning or executing the build.
```

Comment:
An organization that chooses to have their own versioning and dependency system (such as the facebooks and googles of the world) are most likely not going to use Cargo at all. Instead they will have their own tooling that calls rustc directly.

Reply:
A significant motivation for this RFC -- which was designed in coordination with FB and Google build engineers -- is precisely to allow them to manage the build process while still (1) getting access to crates.io and (2) integrating with Rust tooling. Indeed, the sentence you attached this to is specifically talking about how, in these cases, all that Cargo is doing is providing a common way for Rust tooling to get information about a Rust project, even if that information is just being provided by an external build system.